### ISSUE TITLE
Rate Limit Reached

### ISSUE THREAD

**Issue by liw71** [leettools-dev/leettools#14]
hey, I am using deepseek which provides openai compatible api end point. however, I got an error from it, saying 'rate limit reached'. is there an option to reduce the api queries, or have an option to add a delay in each api call? below is full output,
`Mac-mini source % leet flow -t answer -q "how does tcp three way handshake work?" -l info
[30/12/24 12:40:30.121] INFO [2950-131324] Initializing the DuckDB StrategyStore. (strategy_store_duckdb.py:61 __init__)
[30/12/24 12:40:30.168] INFO [2950-131324] Reading extra strategies from path /Volumes/external/Leetools/data/strategy. (strategy_store_duckdb.py:187 _run_init)
[30/12/24 12:40:30.168] INFO [2950-131324] Successfully initialized the DuckDB StrategyStore. (strategy_store_duckdb.py:87 __init__)
[30/12/24 12:40:30.169] INFO [2950-131324] Loaded flow_options: {} (flow_cli.py:290 flow)
[30/12/24 12:40:30.181] INFO [2950-131324] Embedder type for KB adhoc_how_does_20241230_124030 is SegmentEmbedderType.SIMPLE (kb_manager_duckdb.py:131 add_kb)
[30/12/24 12:40:30.181] INFO [2950-131324] Dense Embedder type for KB adhoc_how_does_20241230_124030 is dense_embedder_openai and params {'model_name': 'text-embedding-3-small', 'model_dimension': 1536} (kb_manager_duckdb.py:148 add_kb)
[30/12/24 12:40:30.190] INFO [2950-131324] ChatQueryItem: {'name': 'how does tcp three way handshake work?', 'org_id': 'org-default-id', 'kb_id': '1fe7bfec-02fb-4705-9b36-c8b5834b1a5d', 'creator_id': 'admin', 'article_type': 'chat', 'description': 'Created by CLI command for query how does tcp three way handshake work?', 'share_to_public': False, 'chat_id': 'f528abd0-a281-4b2f-88c2-84b860c832e8', 'owner_id': None, 'created_at': datetime.datetime(2024, 12, 30, 12, 40, 30, 184702), 'updated_at': datetime.datetime(2024, 12, 30, 12, 40, 30, 190403), 'queries': '[{"query_content":"how does tcp three way handshake work?","flow_type":"answer","chat_id":"f528abd0-a281-4b2f-88c2-84b860c832e8","query_id":"bf6a0db9-612b-434e-8ee3-6140f9e1f8db","created_at":"2024-12-30T12:40:30.190403","finished_at":null,"chat_query_options":{"flow_options":{}},"strategy_id":"46de7f7a-66f4-49c4-9eca-8dbaf29ac7ef","strategy_base":null}]', 'answers': '[]', 'metadata': '{}', 'kb_name': None} (history_manager_duckdb.py:487 add_query_item_to_chat)
[30/12/24 12:40:30.193] INFO [Status]Query started.
[30/12/24 12:40:30.194] INFO Searching the web for query: how does tcp three way handshake work?
[30/12/24 12:40:30.194] INFO [Status]Searching the web with RetrieverType.GOOGLE ...
[30/12/24 12:40:30.200] INFO [Status]Start the document process pipeline ...
[30/12/24 12:40:30.221] INFO Searching with query: how does tcp three way handshake work?...
[30/12/24 12:40:30.611] INFO [2950-131324] ✅ Added url to scraper: https://www.geeksforgeeks.org/tcp-3-way-handshake-process/ (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.613] INFO [2950-131324] ✅ Added url to scraper: https://www.sciencedirect.com/topics/computer-science/three-way-handshake (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.615] INFO [2950-131324] ✅ Added url to scraper: https://www.coursera.org/articles/three-way-handshake (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.616] INFO [2950-131324] ✅ Added url to scraper: https://cabulous.medium.com/tcp-3-way-handshake-and-how-it-works-8c5f8d6ea11b (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.618] INFO [2950-131324] ✅ Added url to scraper: https://forum.networklessons.com/t/where-does-the-tcp-3-way-handshake-happen-when-nat-is-involved/4134 (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.619] INFO [2950-131324] ✅ Added url to scraper: https://www.reddit.com/r/networking/comments/8wlow1/is_the_tcp_3way_handshake_always_needed_for/ (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.620] INFO [2950-131324] ✅ Added url to scraper: https://stackoverflow.com/questions/6804979/why-does-tcps-three-way-handshake-bump-the-sequence-number-when-acking (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.621] INFO [2950-131324] ✅ Added url to scraper: https://www.reddit.com/r/explainlikeimfive/comments/6vfjhi/eli5in_networking_what_is_the_three_way_handshake/ (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.623] INFO [2950-131324] ✅ Added url to scraper: https://networkengineering.stackexchange.com/questions/23527/when-does-the-three-way-handshake-take-place-in-relation-to-data-flowing-down-th (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.624] INFO [2950-131324] ✅ Added url to scraper: https://www.reddit.com/r/explainlikeimfive/comments/pawjz9/eli5_tcp_3_way_handshake/ (web_searcher.py:129 _get_new_urls)
[30/12/24 12:40:30.660] INFO 🔍 Trying to scrape: https://www.sciencedirect.com/topics/computer-science/three-way-handshake to dir /Volumes/external/Leetools/data/websearch/sciencedirect
[30/12/24 12:40:30.727] INFO 🔍 Trying to scrape: https://www.reddit.com/r/networking/comments/8wlow1/is_the_tcp_3way_handshake_always_needed_for/ to dir /Volumes/external/Leetools/data/websearch/reddit
[30/12/24 12:40:30.735] INFO 🔍 Trying to scrape: https://www.reddit.com/r/explainlikeimfive/comments/6vfjhi/eli5in_networking_what_is_the_three_way_handshake/ to dir /Volumes/external/Leetools/data/websearch/reddit
[30/12/24 12:40:30.735] INFO 🔄 Reused previously saved data: https://www.reddit.com/r/explainlikeimfive/comments/6vfjhi/eli5in_networking_what_is_the_three_way_handshake/
[30/12/24 12:40:30.735] INFO 🔄 Reused previously saved data: https://www.reddit.com/r/networking/comments/8wlow1/is_the_tcp_3way_handshake_always_needed_for/
[30/12/24 12:40:30.743] INFO 🔍 Trying to scrape: https://networkengineering.stackexchange.com/questions/23527/when-does-the-three-way-handshake-take-place-in-relation-to-data-flowing-down-th to dir /Volumes/external/Leetools/data/websearch/stackexchange
[30/12/24 12:40:30.788] INFO 🔍 Trying to scrape: https://stackoverflow.com/questions/6804979/why-does-tcps-three-way-handshake-bump-the-sequence-number-when-acking to dir /Volumes/external/Leetools/data/websearch/stackoverflow
[30/12/24 12:40:30.798] INFO 🔍 Trying to scrape: https://cabulous.medium.com/tcp-3-way-handshake-and-how-it-works-8c5f8d6ea11b to dir /Volumes/external/Leetools/data/websearch/medium
[30/12/24 12:40:30.807] INFO 🔍 Trying to scrape: https://www.geeksforgeeks.org/tcp-3-way-handshake-process/ to dir /Volumes/external/Leetools/data/websearch/geeksforgeeks
[30/12/24 12:40:30.872] INFO 🔍 Trying to scrape: https://www.reddit.com/r/explainlikeimfive/comments/pawjz9/eli5_tcp_3_way_handshake/ to dir /Volumes/external/Leetools/data/websearch/reddit
[30/12/24 12:40:30.873] INFO 🔄 Reused previously saved data: https://www.geeksforgeeks.org/tcp-3-way-handshake-process/
[30/12/24 12:40:30.873] INFO 🔄 Reused previously saved data: https://www.reddit.com/r/explainlikeimfive/comments/pawjz9/eli5_tcp_3_way_handshake/
[30/12/24 12:40:30.881] INFO 🔍 Trying to scrape: https://www.coursera.org/articles/three-way-handshake to dir /Volumes/external/Leetools/data/websearch/coursera
[30/12/24 12:40:30.938] INFO 🔍 Trying to scrape: https://forum.networklessons.com/t/where-does-the-tcp-3-way-handshake-happen-when-nat-is-involved/4134 to dir /Volumes/external/Leetools/data/websearch/networklessons
[30/12/24 12:40:31.005] INFO 🔄 Reused previously saved data: https://cabulous.medium.com/tcp-3-way-handshake-and-how-it-works-8c5f8d6ea11b
[30/12/24 12:40:31.023] INFO 🔄 Reused previously saved data: https://www.sciencedirect.com/topics/computer-science/three-way-handshake
[30/12/24 12:40:31.058] INFO 🔄 Reused previously saved data: https://networkengineering.stackexchange.com/questions/23527/when-does-the-three-way-handshake-take-place-in-relation-to-data-flowing-down-th
[30/12/24 12:40:31.081] INFO 🔄 Reused previously saved data: https://stackoverflow.com/questions/6804979/why-does-tcps-three-way-handshake-bump-the-sequence-number-when-acking
[30/12/24 12:40:31.340] INFO 🔄 Reused previously saved data: https://www.coursera.org/articles/three-way-handshake
[30/12/24 12:40:31.434] INFO 🔄 Reused previously saved data: https://forum.networklessons.com/t/where-does-the-tcp-3-way-handshake-happen-when-nat-is-involved/4134
[30/12/24 12:40:31.434] INFO Scraped 10 results for how does tcp three way handshake work?
[30/12/24 12:40:31.438] INFO Found 10 docsinks to be created.
[30/12/24 12:40:31.439] INFO Web search job log location: /Volumes/external/Leetools/logs/queries/f528abd0-a281-4b2f-88c2-84b860c832e8/bf6a0db9-612b-434e-8ee3-6140f9e1f8db/web_search_job.log
[30/12/24 12:40:31.439] INFO Adhoc query: converting documents to markdown ...
[30/12/24 12:40:32.657] INFO ✅ Adhoc query: finished converting documents to markdown.
[30/12/24 12:40:32.657] INFO Adhoc query: start to chunk documents ...
[30/12/24 12:40:33.649] INFO ✅ Adhoc query: finished chunking documents.
[30/12/24 12:40:33.649] INFO Adhoc query: start to embed document chunks ...
[30/12/24 12:40:33.808] INFO Adhoc query: got all the chunks 428 ...
[30/12/24 12:40:33.926] INFO [2950-131424] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.927] INFO [2950-131423] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.928] INFO [2950-131430] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.929] INFO [2950-131429] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.929] INFO [2950-131427] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.929] INFO [2950-131425] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.929] INFO [2950-131426] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.930] INFO [2950-131431] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:33.930] INFO [2950-131428] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:39.611] ERROR [2950-131426] OpenAI embedding failed: {
  "event_id": "30-inst-206-20241231044038-2c2fa0da"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:39.627] ERROR [2950-131427] OpenAI embedding failed: {
  "event_id": "30-inst-195-20241231044038-cb924160"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:39.780] ERROR [2950-131428] OpenAI embedding failed: {
  "event_id": "30-inst-128-20241231044039-d53d65ff"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:39.783] ERROR [2950-131423] OpenAI embedding failed: {
  "event_id": "30-inst-33-20241231044039-dd81c358"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:39.834] ERROR [2950-131431] OpenAI embedding failed: {
  "event_id": "30-inst-195-20241231044039-1859cd1d"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:39.915] ERROR [2950-131424] OpenAI embedding failed: {
  "event_id": "30-inst-33-20241231044039-3664cfe3"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:39.932] ERROR [2950-131425] OpenAI embedding failed: {
  "event_id": "30-inst-195-20241231044039-cffc179d"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:40.045] ERROR [2950-131429] OpenAI embedding failed: {
  "event_id": "30-inst-128-20241231044039-72e7317a"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:40.327] ERROR [2950-131430] OpenAI embedding failed: {
  "event_id": "30-inst-128-20241231044039-c0ef29f5"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:40.343] ERROR [2950-131324] Error embedding document: Traceback (most recent call last):
  File "/Volumes/external/source/github/leettools/src/leettools/eds/pipeline/embed/segment_embedder_simple.py", line 63, in embed_segment_list
    rtn_code = self._embed(segments, display_logger)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/eds/pipeline/embed/segment_embedder_simple.py", line 40, in _embed
    op_success = self.dense_vectorstore.save_segments(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/core/repo/_impl/duckdb/vector_store_dense_duckdb.py", line 207, in save_segments
    return self._upsert_embeddings_into_duckdb(org, kb, segments, dense_embedder)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/core/repo/_impl/duckdb/vector_store_dense_duckdb.py", line 181, in _upsert_embeddings_into_duckdb
    rtn_list = list(executor.map(partial_get_embedding, embed_batches))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/concurrent/futures/_base.py", line 619, in result_iterator
    yield _result_or_cancel(fs.pop())
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/concurrent/futures/_base.py", line 317, in _result_or_cancel
    return fut.result(timeout)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/core/repo/_impl/duckdb/vector_store_dense_duckdb.py", line 65, in _batch_get_embedding
    embeddings = self._get_embedding(dense_embedder, texts)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/core/repo/_impl/duckdb/vector_store_dense_duckdb.py", line 128, in _get_embedding
    embed_result: DenseEmbeddings = dense_embedder.embed(embed_request)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/eds/str_embedder/_impl/dense_embedder_openai.py", line 101, in embed
    raise e
  File "/Volumes/external/source/github/leettools/src/leettools/eds/str_embedder/_impl/dense_embedder_openai.py", line 93, in embed
    response = openai.embeddings.create(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/resources/embeddings.py", line 124, in create
    return self._post(
           ^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1280, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 957, in request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1046, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1095, in _retry_request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1046, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1095, in _retry_request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1061, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.RateLimitError: {
  "event_id": "30-inst-33-20241231044039-dd81c358"
  "error_msg": "Rate Limit Reached."
}
 (segment_embedder_simple.py:67 embed_segment_list)
[30/12/24 12:40:40.343] INFO Adhoc query: embedded all the chunks ...
[30/12/24 12:40:40.423] INFO ✅ Adhoc query: finished embedding documents chunks.
[30/12/24 12:40:40.426] INFO Successfully processed 10 documents.
[30/12/24 12:40:40.426] INFO [Status]Search in KB adhoc_how_does_20241230_124030 for related segments.
[30/12/24 12:40:40.430] INFO [2950-131324] The filter expression is: None for query how does tcp three way handshake work? (searcher_simple.py:53 execute_kb_search)
[30/12/24 12:40:40.442] INFO [2950-131324] Creating OpenAI embedding client with base_url: https://api.deepseek.com/v1 and api_key: sk-af******462d8 (api_utils.py:222 get_openai_embedder_client_for_user)
[30/12/24 12:40:45.249] ERROR [2950-131324] OpenAI embedding failed: {
  "event_id": "30-inst-33-20241231044044-5f530ddb"
  "error_msg": "Rate Limit Reached."
} (dense_embedder_openai.py:100 embed)
[30/12/24 12:40:45.258] ERROR Error in getting answer for query: {
  "event_id": "30-inst-33-20241231044044-5f530ddb"
  "error_msg": "Rate Limit Reached."
}

 Traceback (most recent call last):
  File "/Volumes/external/source/github/leettools/src/leettools/chat/_impl/duckdb/history_manager_duckdb.py", line 303, in _execute_flow_for_query
    return flow.execute_query(
           ^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/flow/flows/answer/flow_answer.py", line 122, in execute_query
    top_ranked_result_segments = steps.StepVectorSearch.run_step(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/flow/steps/step_vectdb_search.py", line 163, in run_step
    top_ranked_result_segments = searcher.execute_kb_search(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/eds/rag/search/_impl/searcher_simple.py", line 54, in execute_kb_search
    results: List[VectorSearchResult] = self.dense_vectorstore.search_in_kb(
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/core/repo/_impl/duckdb/vector_store_dense_duckdb.py", line 306, in search_in_kb
    embed_result: DenseEmbeddings = dense_embedder.embed(embed_request)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/external/source/github/leettools/src/leettools/eds/str_embedder/_impl/dense_embedder_openai.py", line 101, in embed
    raise e
  File "/Volumes/external/source/github/leettools/src/leettools/eds/str_embedder/_impl/dense_embedder_openai.py", line 93, in embed
    response = openai.embeddings.create(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/resources/embeddings.py", line 124, in create
    return self._post(
           ^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1280, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 957, in request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1046, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1095, in _retry_request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1046, in _request
    return self._retry_request(
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1095, in _retry_request
    return self._request(
           ^^^^^^^^^^^^^^
  File "/Users/weili/anaconda3/envs/leettools/lib/python3.11/site-packages/openai/_base_client.py", line 1061, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.RateLimitError: {
  "event_id": "30-inst-33-20241231044044-5f530ddb"
  "error_msg": "Rate Limit Reached."
}

[30/12/24 12:40:45.259] INFO [Status]Saving results.
[30/12/24 12:40:45.264] INFO [Status]Query completed.
# How Does Tcp Three Way Handshake Work?
`

**Comment by pengfeng** [comment]
Thanks for reporting the issue!

The problem is caused by the embedding model: when none is specified, it will use the OpenAI embedding API and share the endpoint with the inference API. I don't this approach works for the DeepSeek API.

I am not sure which embedding model you usually use or if they are compatible with OpenAI? We can see if we can add the support for those APIs.

Or if you do not mind use a local simple embedder, you can use 

`EDS_DEFAULT_DENSE_EMBEDDER="dense_embedder_sentence_transformer"  leet flow -t answer -q "how does tcp three way handshake work?" -l info`

**Comment by liw71** [comment]
just found out deepseek does not support embedding api yet, https://github.com/deepseek-ai/DeepSeek-V2/issues/67
their pricing is much cheaper, hopefully they will support it soon.

thanks much again for quick response.

**Comment by pengfeng** [comment]
We now added support for DeepSeek API, just set 

 export EDS_DEFAULT_OPENAI_BASE_URL=https://api.deepseek.com/v1
 export EDS_OPENAI_API_KEY=<your deepseek API key>
 # this one will use a local embedder since deepseek does not provide embedder model yet
 export EDS_DEFAULT_DENSE_EMBEDDER=dense_embedder_local_mem
 export EDS_DEFAULT_OPENAI_MODEL=deepseek-chat
 
 And all the inferences will be using DeepSeek endpoints.
